File: src/layout.cpp
#include "hanfe/layout.hpp"

#include <linux/input-event-codes.h>

#include <algorithm>
#include <stdexcept>

namespace hanfe {
namespace {

LayoutSymbol make_text_symbol(const std::string& value, bool commit_before = true) {
    LayoutSymbol symbol;
    symbol.kind = SymbolKind::Text;
    symbol.text = value;
    symbol.commit_before = commit_before;
    return symbol;
}

LayoutSymbol make_jamo_symbol(char32_t value, JamoRole role = JamoRole::Auto) {
    LayoutSymbol symbol;
    symbol.kind = SymbolKind::Jamo;
    symbol.jamo = value;
    symbol.role = role;
    symbol.commit_before = false;
    return symbol;
}

LayoutSymbol make_passthrough_symbol(bool commit_before = true) {
    LayoutSymbol symbol;
    symbol.kind = SymbolKind::Passthrough;
    symbol.commit_before = commit_before;
    return symbol;
}

Layout build_dubeolsik() {
    std::unordered_map<int, LayoutEntry> mapping;

    auto add = [&](int key, LayoutSymbol normal, std::optional<LayoutSymbol> shifted = std::nullopt) {
        LayoutEntry entry;
        entry.normal = normal;
        entry.shifted = shifted;
        mapping.emplace(key, entry);
    };

    add(KEY_Q, make_jamo_symbol(U'ㅂ'), make_jamo_symbol(U'ㅃ'));
    add(KEY_W, make_jamo_symbol(U'ㅈ'), make_jamo_symbol(U'ㅉ'));
    add(KEY_E, make_jamo_symbol(U'ㄷ'), make_jamo_symbol(U'ㄸ'));
    add(KEY_R, make_jamo_symbol(U'ㄱ'), make_jamo_symbol(U'ㄲ'));
    add(KEY_T, make_jamo_symbol(U'ㅅ'), make_jamo_symbol(U'ㅆ'));
    add(KEY_Y, make_jamo_symbol(U'ㅛ'));
    add(KEY_U, make_jamo_symbol(U'ㅕ'));
    add(KEY_I, make_jamo_symbol(U'ㅑ'));
    add(KEY_O, make_jamo_symbol(U'ㅐ'), make_jamo_symbol(U'ㅒ'));
    add(KEY_P, make_jamo_symbol(U'ㅔ'), make_jamo_symbol(U'ㅖ'));
    add(KEY_A, make_jamo_symbol(U'ㅁ'));
    add(KEY_S, make_jamo_symbol(U'ㄴ'));
    add(KEY_D, make_jamo_symbol(U'ㅇ'));
    add(KEY_F, make_jamo_symbol(U'ㄹ'));
    add(KEY_G, make_jamo_symbol(U'ㅎ'));
    add(KEY_H, make_jamo_symbol(U'ㅗ'));
    add(KEY_J, make_jamo_symbol(U'ㅓ'));
    add(KEY_K, make_jamo_symbol(U'ㅏ'));
    add(KEY_L, make_jamo_symbol(U'ㅣ'));

    {
        LayoutEntry entry;
        entry.normal = make_text_symbol(";");
        entry.shifted = make_text_symbol(":");
        mapping.emplace(KEY_SEMICOLON, entry);
    }
    {
        LayoutEntry entry;
        entry.normal = make_text_symbol("'");
        entry.shifted = make_text_symbol("\"");
        mapping.emplace(KEY_APOSTROPHE, entry);
    }

    add(KEY_Z, make_jamo_symbol(U'ㅋ'));
    add(KEY_X, make_jamo_symbol(U'ㅌ'));
    add(KEY_C, make_jamo_symbol(U'ㅊ'));
    add(KEY_V, make_jamo_symbol(U'ㅍ'));
    add(KEY_B, make_jamo_symbol(U'ㅠ'));
    add(KEY_N, make_jamo_symbol(U'ㅜ'));
    add(KEY_M, make_jamo_symbol(U'ㅡ'));

    {
        LayoutEntry entry;
        entry.normal = make_text_symbol(",");
        entry.shifted = make_text_symbol("<");
        mapping.emplace(KEY_COMMA, entry);
    }
    {
        LayoutEntry entry;
        entry.normal = make_text_symbol(".");
        entry.shifted = make_text_symbol(">");
        mapping.emplace(KEY_DOT, entry);
    }
    {
        LayoutEntry entry;
        entry.normal = make_text_symbol("/");
        entry.shifted = make_text_symbol("?");
        mapping.emplace(KEY_SLASH, entry);
    }

    add(KEY_SPACE, make_text_symbol(" "));

    auto add_number = [&](int key, const std::string& normal, const std::string& shifted) {
        LayoutEntry entry;
        entry.normal = make_text_symbol(normal);
        entry.shifted = make_text_symbol(shifted);
        mapping.emplace(key, entry);
    };

    add_number(KEY_1, "1", "!");
    add_number(KEY_2, "2", "@");
    add_number(KEY_3, "3", "#");
    add_number(KEY_4, "4", "$");
    add_number(KEY_5, "5", "%");
    add_number(KEY_6, "6", "^");
    add_number(KEY_7, "7", "&");
    add_number(KEY_8, "8", "*");
    add_number(KEY_9, "9", "(");
    add_number(KEY_0, "0", ")");
    add_number(KEY_MINUS, "-", "_");
    add_number(KEY_EQUAL, "=", "+");
    add_number(KEY_LEFTBRACE, "[", "{");
    add_number(KEY_RIGHTBRACE, "]", "}");

    {
        LayoutEntry entry;
        entry.normal = make_text_symbol("\\");
        entry.shifted = make_text_symbol("|");
        mapping.emplace(KEY_BACKSLASH, entry);
    }
    {
        LayoutEntry entry;
        entry.normal = make_text_symbol("`");
        entry.shifted = make_text_symbol("~");
        mapping.emplace(KEY_GRAVE, entry);
    }

    mapping.emplace(KEY_TAB, LayoutEntry{make_passthrough_symbol(), std::nullopt});
    mapping.emplace(KEY_ENTER, LayoutEntry{make_passthrough_symbol(), std::nullopt});
    mapping.emplace(KEY_ESC, LayoutEntry{make_passthrough_symbol(), std::nullopt});
    mapping.emplace(KEY_BACKSPACE, LayoutEntry{make_passthrough_symbol(), std::nullopt});

    return Layout("dubeolsik", std::move(mapping));
}

Layout build_sebeolsik_390() {
    std::unordered_map<int, LayoutEntry> mapping;

    auto set_entry = [&](int key, const LayoutSymbol& normal, std::optional<LayoutSymbol> shifted) {
        LayoutEntry entry;
        entry.normal = normal;
        entry.shifted = shifted;
        mapping.emplace(key, entry);
    };

    auto text = [](const std::string& value) { return make_text_symbol(value); };
    auto jamo = [](char32_t value) { return make_jamo_symbol(value); };

    set_entry(KEY_GRAVE, text("`"), make_text_symbol("~"));
    set_entry(KEY_1, text("1"), make_text_symbol("!"));
    set_entry(KEY_2, text("2"), make_text_symbol("@"));
    set_entry(KEY_3, text("3"), make_text_symbol("#"));
    set_entry(KEY_4, text("4"), make_text_symbol("$"));
    set_entry(KEY_5, text("5"), make_text_symbol("%"));
    set_entry(KEY_6, text("6"), make_text_symbol("^"));
    set_entry(KEY_7, text("7"), make_text_symbol("&"));
    set_entry(KEY_8, text("8"), make_text_symbol("*"));
    set_entry(KEY_9, text("9"), make_text_symbol("("));
    set_entry(KEY_0, text("0"), make_text_symbol(")"));
    set_entry(KEY_MINUS, text("-"), make_text_symbol("_"));
    set_entry(KEY_EQUAL, text("="), make_text_symbol("+"));

    set_entry(KEY_Q, jamo(U'ㅂ'), make_jamo_symbol(U'ㅃ'));
    set_entry(KEY_W, jamo(U'ㅈ'), make_jamo_symbol(U'ㅉ'));
    set_entry(KEY_E, jamo(U'ㄷ'), make_jamo_symbol(U'ㄸ'));
    set_entry(KEY_R, jamo(U'ㄱ'), make_jamo_symbol(U'ㄲ'));
    set_entry(KEY_T, jamo(U'ㅅ'), make_jamo_symbol(U'ㅆ'));
    set_entry(KEY_Y, jamo(U'ㅛ'), make_jamo_symbol(U'ㅅ', JamoRole::Trailing));
    set_entry(KEY_U, jamo(U'ㅕ'), make_jamo_symbol(U'ㅈ', JamoRole::Trailing));
    set_entry(KEY_I, jamo(U'ㅑ'), make_jamo_symbol(U'ㅊ', JamoRole::Trailing));
    set_entry(KEY_O, jamo(U'ㅐ'), make_jamo_symbol(U'ㅋ', JamoRole::Trailing));
    set_entry(KEY_P, jamo(U'ㅔ'), make_jamo_symbol(U'ㅌ', JamoRole::Trailing));
    set_entry(KEY_LEFTBRACE, jamo(U'ㅒ'), make_jamo_symbol(U'ㅍ', JamoRole::Trailing));
    set_entry(KEY_RIGHTBRACE, jamo(U'ㅖ'), make_jamo_symbol(U'ㅎ', JamoRole::Trailing));
    {
        LayoutEntry entry;
        entry.normal = jamo(U'ㅢ');
        entry.shifted = make_text_symbol("|");
        mapping.emplace(KEY_BACKSLASH, entry);
    }

    set_entry(KEY_A, jamo(U'ㅁ'), make_jamo_symbol(U'ㅁ'));
    set_entry(KEY_S, jamo(U'ㄴ'), make_jamo_symbol(U'ㄴ'));
    set_entry(KEY_D, jamo(U'ㅇ'), make_jamo_symbol(U'ㅇ'));
    set_entry(KEY_F, jamo(U'ㄹ'), make_jamo_symbol(U'ㄹ'));
    set_entry(KEY_G, jamo(U'ㅎ'), make_jamo_symbol(U'ㅎ'));
    set_entry(KEY_H, jamo(U'ㅗ'), make_jamo_symbol(U'ㄱ', JamoRole::Trailing));
    set_entry(KEY_J, jamo(U'ㅓ'), make_jamo_symbol(U'ㄴ', JamoRole::Trailing));
    set_entry(KEY_K, jamo(U'ㅏ'), make_jamo_symbol(U'ㄷ', JamoRole::Trailing));
    set_entry(KEY_L, jamo(U'ㅣ'), make_jamo_symbol(U'ㄹ', JamoRole::Trailing));
    set_entry(KEY_SEMICOLON, jamo(U'ㅠ'), make_jamo_symbol(U'ㅁ', JamoRole::Trailing));
    set_entry(KEY_APOSTROPHE, jamo(U'ㅜ'), make_jamo_symbol(U'ㅂ', JamoRole::Trailing));

    set_entry(KEY_Z, jamo(U'ㅋ'), make_jamo_symbol(U'ㅋ'));
    set_entry(KEY_X, jamo(U'ㅌ'), make_jamo_symbol(U'ㅌ'));
    set_entry(KEY_C, jamo(U'ㅊ'), make_jamo_symbol(U'ㅊ'));
    set_entry(KEY_V, jamo(U'ㅍ'), make_jamo_symbol(U'ㅍ'));
    set_entry(KEY_B, jamo(U'ㅠ'), make_jamo_symbol(U'ㅇ', JamoRole::Trailing));
    set_entry(KEY_N, jamo(U'ㅜ'), make_jamo_symbol(U'ㅅ', JamoRole::Trailing));
    set_entry(KEY_M, jamo(U'ㅡ'), make_jamo_symbol(U'ㅎ', JamoRole::Trailing));

    {
        LayoutEntry entry;
        entry.normal = jamo(U'ㅘ');
        entry.shifted = jamo(U'ㅙ');
        mapping.emplace(KEY_COMMA, entry);
    }
    {
        LayoutEntry entry;
        entry.normal = jamo(U'ㅝ');
        entry.shifted = jamo(U'ㅞ');
        mapping.emplace(KEY_DOT, entry);
    }
    {
        LayoutEntry entry;
        entry.normal = jamo(U'ㅟ');
        entry.shifted = make_text_symbol("?");
        mapping.emplace(KEY_SLASH, entry);
    }

    set_entry(KEY_SPACE, text(" "), std::nullopt);

    mapping.emplace(KEY_ENTER, LayoutEntry{make_passthrough_symbol(), std::nullopt});
    mapping.emplace(KEY_TAB, LayoutEntry{make_passthrough_symbol(), std::nullopt});
    mapping.emplace(KEY_ESC, LayoutEntry{make_passthrough_symbol(), std::nullopt});
    mapping.emplace(KEY_BACKSPACE, LayoutEntry{make_passthrough_symbol(), std::nullopt});

    return Layout("sebeolsik-390", std::move(mapping));
}

}  // namespace

Layout::Layout(std::string name, std::unordered_map<int, LayoutEntry> mapping)
    : name_(std::move(name)), mapping_(std::move(mapping)) {}

const LayoutSymbol* Layout::translate(int keycode, bool shift_active) const {
    auto it = mapping_.find(keycode);
    if (it == mapping_.end()) {
        return nullptr;
    }
    const LayoutEntry& entry = it->second;
    if (shift_active && entry.shifted) {
        return &entry.shifted.value();
    }
    if (entry.normal) {
        return &entry.normal.value();
    }
    if (entry.shifted) {
        return &entry.shifted.value();
    }
    return nullptr;
}

Layout make_dubeolsik_layout() { return build_dubeolsik(); }

Layout make_sebeolsik_390_layout() { return build_sebeolsik_390(); }

std::vector<std::string> available_layouts() {
    std::vector<std::string> names = {"dubeolsik", "sebeolsik-390"};
    std::sort(names.begin(), names.end());
    return names;
}

Layout load_layout(const std::string& name) {
    if (name == "dubeolsik") {
        return make_dubeolsik_layout();
    }
    if (name == "sebeolsik-390") {
        return make_sebeolsik_390_layout();
    }
    throw std::runtime_error("Unknown layout: " + name);
}

std::unordered_map<char, int> unicode_hex_keycodes() {
    return {
        {'0', KEY_0}, {'1', KEY_1}, {'2', KEY_2}, {'3', KEY_3}, {'4', KEY_4}, {'5', KEY_5},
        {'6', KEY_6}, {'7', KEY_7}, {'8', KEY_8}, {'9', KEY_9}, {'a', KEY_A}, {'b', KEY_B},
        {'c', KEY_C}, {'d', KEY_D}, {'e', KEY_E}, {'f', KEY_F}};
}

}  // namespace hanfe

File: internal/layout/layout.go
package layout

import (
	"fmt"
	"sort"

	"github.com/gg582/hanfe/internal/hangul"
	"github.com/gg582/hanfe/internal/linux"
)

type SymbolKind int

const (
	SymbolPassthrough SymbolKind = iota
	SymbolText
	SymbolJamo
)

type LayoutSymbol struct {
	Kind         SymbolKind
	Text         string
	Jamo         rune
	Role         hangul.JamoRole
	CommitBefore bool
}

type LayoutEntry struct {
	Normal  *LayoutSymbol
	Shifted *LayoutSymbol
}

type Category string

const (
	CategoryHangul Category = "hangul"
	CategoryKana   Category = "kana"
)

type Layout struct {
	name     string
	category Category
	mapping  map[uint16]LayoutEntry
}

func (l Layout) Name() string { return l.name }

func (l Layout) Category() Category { return l.category }

func (l Layout) Translate(code uint16, shift bool) *LayoutSymbol {
	entry, ok := l.mapping[code]
	if !ok {
		return nil
	}
	if shift && entry.Shifted != nil {
		return entry.Shifted
	}
	if entry.Normal != nil {
		return entry.Normal
	}
	if entry.Shifted != nil {
		return entry.Shifted
	}
	return nil
}

func makeJamoSymbol(value rune, role hangul.JamoRole) *LayoutSymbol {
	return &LayoutSymbol{Kind: SymbolJamo, Jamo: value, Role: role}
}

func makeJamo(value rune) *LayoutSymbol { return makeJamoSymbol(value, hangul.RoleAuto) }

func makePassthroughSymbol(commitBefore bool) *LayoutSymbol {
	return &LayoutSymbol{Kind: SymbolPassthrough, CommitBefore: commitBefore}
}

func addEntry(mapping map[uint16]LayoutEntry, key int, normal *LayoutSymbol, shifted *LayoutSymbol) {
	mapping[uint16(key)] = LayoutEntry{Normal: normal, Shifted: shifted}
}

func buildDubeolsik() Layout {
	mapping := make(map[uint16]LayoutEntry)
	addEntry(mapping, linux.KeyQ, makeJamo('ㅂ'), makeJamo('ㅃ'))
	addEntry(mapping, linux.KeyW, makeJamo('ㅈ'), makeJamo('ㅉ'))
	addEntry(mapping, linux.KeyE, makeJamo('ㄷ'), makeJamo('ㄸ'))
	addEntry(mapping, linux.KeyR, makeJamo('ㄱ'), makeJamo('ㄲ'))
	addEntry(mapping, linux.KeyT, makeJamo('ㅅ'), makeJamo('ㅆ'))
	addEntry(mapping, linux.KeyY, makeJamo('ㅛ'), nil)
	addEntry(mapping, linux.KeyU, makeJamo('ㅕ'), nil)
	addEntry(mapping, linux.KeyI, makeJamo('ㅑ'), nil)
	addEntry(mapping, linux.KeyO, makeJamo('ㅐ'), makeJamo('ㅒ'))
	addEntry(mapping, linux.KeyP, makeJamo('ㅔ'), makeJamo('ㅖ'))
	addEntry(mapping, linux.KeyA, makeJamo('ㅁ'), nil)
	addEntry(mapping, linux.KeyS, makeJamo('ㄴ'), nil)
	addEntry(mapping, linux.KeyD, makeJamo('ㅇ'), nil)
	addEntry(mapping, linux.KeyF, makeJamo('ㄹ'), nil)
	addEntry(mapping, linux.KeyG, makeJamo('ㅎ'), nil)
	addEntry(mapping, linux.KeyH, makeJamo('ㅗ'), nil)
	addEntry(mapping, linux.KeyJ, makeJamo('ㅓ'), nil)
	addEntry(mapping, linux.KeyK, makeJamo('ㅏ'), nil)
	addEntry(mapping, linux.KeyL, makeJamo('ㅣ'), nil)

	passthrough := makePassthroughSymbol(true)
	addEntry(mapping, linux.KeySemicolon, passthrough, passthrough)
	addEntry(mapping, linux.KeyApostrophe, passthrough, passthrough)

	addEntry(mapping, linux.KeyZ, makeJamo('ㅋ'), nil)
	addEntry(mapping, linux.KeyX, makeJamo('ㅌ'), nil)
	addEntry(mapping, linux.KeyC, makeJamo('ㅊ'), nil)
	addEntry(mapping, linux.KeyV, makeJamo('ㅍ'), nil)
	addEntry(mapping, linux.KeyB, makeJamo('ㅠ'), nil)
	addEntry(mapping, linux.KeyN, makeJamo('ㅜ'), nil)
	addEntry(mapping, linux.KeyM, makeJamo('ㅡ'), nil)

	addEntry(mapping, linux.KeyComma, passthrough, passthrough)
	addEntry(mapping, linux.KeyDot, passthrough, passthrough)
	addEntry(mapping, linux.KeySlash, passthrough, passthrough)

	addEntry(mapping, linux.KeySpace, makePassthroughSymbol(true), nil)

	passthroughPairKeys := []int{
		linux.Key1, linux.Key2, linux.Key3, linux.Key4, linux.Key5,
		linux.Key6, linux.Key7, linux.Key8, linux.Key9, linux.Key0,
		linux.KeyMinus, linux.KeyEqual, linux.KeyLeftBrace, linux.KeyRightBrace,
		linux.KeyBackslash, linux.KeyGrave,
	}
	for _, key := range passthroughPairKeys {
		addEntry(mapping, key, passthrough, passthrough)
	}

	specialKeys := []int{linux.KeyTab, linux.KeyEnter, linux.KeyEsc, linux.KeyBackspace}
	for _, key := range specialKeys {
		addEntry(mapping, key, makePassthroughSymbol(true), nil)
	}

	return Layout{name: "dubeolsik", category: CategoryHangul, mapping: mapping}
}

func buildSebeolsik390() Layout {
	mapping := make(map[uint16]LayoutEntry)
	passthrough := makePassthroughSymbol(true)

	passthroughKeys := []int{
		linux.KeyGrave,
		linux.Key1, linux.Key2, linux.Key3, linux.Key4, linux.Key5,
		linux.Key6, linux.Key7, linux.Key8, linux.Key9, linux.Key0,
		linux.KeyMinus, linux.KeyEqual,
	}
	for _, key := range passthroughKeys {
		addEntry(mapping, key, passthrough, passthrough)
	}

	addEntry(mapping, linux.KeyQ, makeJamo('ㅂ'), makeJamo('ㅃ'))
	addEntry(mapping, linux.KeyW, makeJamo('ㅈ'), makeJamo('ㅉ'))
	addEntry(mapping, linux.KeyE, makeJamo('ㄷ'), makeJamo('ㄸ'))
	addEntry(mapping, linux.KeyR, makeJamo('ㄱ'), makeJamo('ㄲ'))
	addEntry(mapping, linux.KeyT, makeJamo('ㅅ'), makeJamo('ㅆ'))
	addEntry(mapping, linux.KeyY, makeJamo('ㅛ'), makeJamoSymbol('ㅅ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyU, makeJamo('ㅕ'), makeJamoSymbol('ㅈ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyI, makeJamo('ㅑ'), makeJamoSymbol('ㅊ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyO, makeJamo('ㅐ'), makeJamoSymbol('ㅋ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyP, makeJamo('ㅔ'), makeJamoSymbol('ㅌ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyLeftBrace, makeJamo('ㅒ'), makeJamoSymbol('ㅍ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyRightBrace, makeJamo('ㅖ'), makeJamoSymbol('ㅎ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyBackslash, makeJamo('ㅢ'), passthrough)

	addEntry(mapping, linux.KeyA, makeJamo('ㅁ'), makeJamo('ㅁ'))
	addEntry(mapping, linux.KeyS, makeJamo('ㄴ'), makeJamo('ㄴ'))
	addEntry(mapping, linux.KeyD, makeJamo('ㅇ'), makeJamo('ㅇ'))
	addEntry(mapping, linux.KeyF, makeJamo('ㄹ'), makeJamo('ㄹ'))
	addEntry(mapping, linux.KeyG, makeJamo('ㅎ'), makeJamo('ㅎ'))
	addEntry(mapping, linux.KeyH, makeJamo('ㅗ'), makeJamoSymbol('ㄱ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyJ, makeJamo('ㅓ'), makeJamoSymbol('ㄴ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyK, makeJamo('ㅏ'), makeJamoSymbol('ㄷ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyL, makeJamo('ㅣ'), makeJamoSymbol('ㄹ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeySemicolon, makeJamo('ㅠ'), makeJamoSymbol('ㅁ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyApostrophe, makeJamo('ㅜ'), makeJamoSymbol('ㅂ', hangul.RoleTrailing))

	addEntry(mapping, linux.KeyZ, makeJamo('ㅋ'), makeJamo('ㅋ'))
	addEntry(mapping, linux.KeyX, makeJamo('ㅌ'), makeJamo('ㅌ'))
	addEntry(mapping, linux.KeyC, makeJamo('ㅊ'), makeJamo('ㅊ'))
	addEntry(mapping, linux.KeyV, makeJamo('ㅍ'), makeJamo('ㅍ'))
	addEntry(mapping, linux.KeyB, makeJamo('ㅠ'), makeJamoSymbol('ㅇ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyN, makeJamo('ㅜ'), makeJamoSymbol('ㅅ', hangul.RoleTrailing))
	addEntry(mapping, linux.KeyM, makeJamo('ㅡ'), makeJamoSymbol('ㅎ', hangul.RoleTrailing))

	addEntry(mapping, linux.KeyComma, makeJamo('ㅘ'), makeJamo('ㅙ'))
	addEntry(mapping, linux.KeyDot, makeJamo('ㅝ'), makeJamo('ㅞ'))
	addEntry(mapping, linux.KeySlash, makeJamo('ㅟ'), passthrough)

	addEntry(mapping, linux.KeySpace, makePassthroughSymbol(true), nil)

	specialKeys := []int{linux.KeyEnter, linux.KeyTab, linux.KeyEsc, linux.KeyBackspace}
	for _, key := range specialKeys {
		addEntry(mapping, key, makePassthroughSymbol(true), nil)
	}

	return Layout{name: "sebeolsik-390", category: CategoryHangul, mapping: mapping}
}

func makeTextSymbol(value string, commitBefore bool) *LayoutSymbol {
	return &LayoutSymbol{Kind: SymbolText, Text: value, CommitBefore: commitBefore}
}

func hiraganaToKatakana(r rune) rune {
	// Hiragana and Katakana blocks are offset by a constant value.
	const hiraganaStart = rune(0x3041)
	const hiraganaEnd = rune(0x3096)
	const katakanaStart = rune(0x30A1)
	offset := katakanaStart - hiraganaStart
	if r >= hiraganaStart && r <= hiraganaEnd {
		return r + offset
	}
	return r
}

func hiraganaPair(normal string) (string, string) {
	if normal == "" {
		return "", ""
	}
	runes := []rune(normal)
	shifted := make([]rune, len(runes))
	for i, r := range runes {
		shifted[i] = hiraganaToKatakana(r)
	}
	return normal, string(shifted)
}

func buildKana86() Layout {
	mapping := make(map[uint16]LayoutEntry)

	addKana := func(key int, normal string) {
		n, s := hiraganaPair(normal)
		addEntry(mapping, key, makeTextSymbol(n, false), makeTextSymbol(s, false))
	}

	addKana(linux.Key1, "ぬ")
	addKana(linux.Key2, "ふ")
	addKana(linux.Key3, "あ")
	addKana(linux.Key4, "う")
	addKana(linux.Key5, "え")
	addKana(linux.Key6, "お")
	addKana(linux.Key7, "や")
	addKana(linux.Key8, "ゆ")
	addKana(linux.Key9, "よ")
	addKana(linux.Key0, "わ")
	addKana(linux.KeyMinus, "ほ")
	addKana(linux.KeyEqual, "へ")

	addKana(linux.KeyQ, "た")
	addKana(linux.KeyW, "て")
	addKana(linux.KeyE, "い")
	addKana(linux.KeyR, "す")
	addKana(linux.KeyT, "か")
	addKana(linux.KeyY, "ん")
	addKana(linux.KeyU, "な")
	addKana(linux.KeyI, "に")
	addKana(linux.KeyO, "ら")
	addKana(linux.KeyP, "せ")
	addKana(linux.KeyLeftBrace, "゛")
	addKana(linux.KeyRightBrace, "゜")

	addKana(linux.KeyA, "ち")
	addKana(linux.KeyS, "と")
	addKana(linux.KeyD, "し")
	addKana(linux.KeyF, "は")
	addKana(linux.KeyG, "き")
	addKana(linux.KeyH, "く")
	addKana(linux.KeyJ, "ま")
	addKana(linux.KeyK, "の")
	addKana(linux.KeyL, "り")
	addKana(linux.KeySemicolon, "れ")
	addKana(linux.KeyApostrophe, "け")
	addKana(linux.KeyBackslash, "む")

	addKana(linux.KeyZ, "つ")
	addKana(linux.KeyX, "さ")
	addKana(linux.KeyC, "そ")
	addKana(linux.KeyV, "ひ")
	addKana(linux.KeyB, "こ")
	addKana(linux.KeyN, "み")
	addKana(linux.KeyM, "も")
	addKana(linux.KeyComma, "ね")
	addKana(linux.KeyDot, "る")
	addKana(linux.KeySlash, "め")

	space := makePassthroughSymbol(true)
	addEntry(mapping, linux.KeySpace, space, nil)

	passthroughKeys := []int{linux.KeyTab, linux.KeyEnter, linux.KeyEsc, linux.KeyBackspace}
	for _, key := range passthroughKeys {
		addEntry(mapping, key, space, nil)
	}

	return Layout{name: "kana86", category: CategoryKana, mapping: mapping}
}

func AvailableLayouts() []string {
	names := []string{"dubeolsik", "kana86", "sebeolsik-390"}
	sort.Strings(names)
	return names
}

func Load(name string) (Layout, error) {
	switch name {
	case "", "dubeolsik":
		return buildDubeolsik(), nil
	case "sebeolsik-390":
		return buildSebeolsik390(), nil
	case "kana86":
		return buildKana86(), nil
	default:
		return Layout{}, fmt.Errorf("unknown layout: %s", name)
	}
}

func UnicodeHexKeycodes() map[rune]uint16 {
	return map[rune]uint16{
		'0': uint16(linux.Key0),
		'1': uint16(linux.Key1),
		'2': uint16(linux.Key2),
		'3': uint16(linux.Key3),
		'4': uint16(linux.Key4),
		'5': uint16(linux.Key5),
		'6': uint16(linux.Key6),
		'7': uint16(linux.Key7),
		'8': uint16(linux.Key8),
		'9': uint16(linux.Key9),
		'a': uint16(linux.KeyA),
		'b': uint16(linux.KeyB),
		'c': uint16(linux.KeyC),
		'd': uint16(linux.KeyD),
		'e': uint16(linux.KeyE),
		'f': uint16(linux.KeyF),
	}
}
